design-documents/mail-library-improvements.md

```diff
@@ -0,0 +1,338 @@
+# Mail library improvements
+
+## About
+
+Existing `\Magento\Framework\Mail\MailMessageInterface` missing many essential methods. There is no way to get data set on the object and library doesn't support sending attachments. This document describes improvements to Mail library.
+
+## Design
+
+Improvements to the library can be done in the following stages
+1. Add missing methods to the `\Magento\Framework\Mail\MailMessageInterface` interface
+2. Introduce new library with more clear interface that would allow to send attachments (wrappers around Zend\Mail)
+3. Refactor Magento modules to use new library
+
+1. Improving `\Magento\Framework\Mail\MailMessageInterface`
+    Below is the comparison with `Zend\Mail\Message`.
+    
+    - `isValid` - does’t need to be part of interface
+    + `setEncoding`/`getEncoding` - can pass in constructor
+    - `setHeaders`/`getHeaders`
+    + `setFrom`
+    - `getFrom`
+    - `setTo` - there is addCc, but not convenient
```

Comment:
As `\Magento\Framework\Mail\MailMessageInterface` extends `\Magento\Framework\Mail\MessageInterface`, which already has `addCc`, does it make sense to introduce a new one method?

Reply:
I would introduce it, since `addCc` doesn't allow to replace.